### ISSUE TITLE
Cache filters

### ISSUE THREAD

**Issue by szuecs** [zalando/skipper#3150]
**Is your feature request related to a problem? Please describe.**

Caching API requests for some time defined.

**Describe the solution you would like**

I imagine `cache("lru", "1m")` to cache for a given request the response for 1 minute and respond with the same headers and body as the original response. It would use "lru" as eviction policy.
It should somehow work similar to more solid cache servers like [Varnish](https://varnish-cache.org/intro/index.html#intro).

I think we should support multiple algorithms to do cache eviction https://github.com/zalando/skipper/issues/956
Later we can also support a groupcache to serve with all skipper fleet so having a huge cache size as memory and use something like consistent hash to find the location of the data https://github.com/zalando/skipper/issues/628.

We can re-use the valkey ring shards implementation and use GET/SET operations to have a large "group cache" like behavior.

references:
- RFC: https://datatracker.ietf.org/doc/html/rfc7234
- CloudFront cache behavior: https://docs.aws.amazon.com/cloudfront/latest/APIReference/API_CacheBehavior.html

**Describe alternatives you've considered (optional)**

CDN or static blob storage (e.g. S3), but both are not dynamic.
[Varnish](https://varnish-cache.org/intro/index.html#intro) has nice caching capabilities.

**Comment by AlexanderYastrebov** [comment]
We have Redis integration so maybe we can implement it using redis commands or script (think `redis` filter).

**Comment by szuecs** [comment]
Yes maybe, but this issue does not say anything about implementation specifics, so could be redis but does not have to.
In our case it likely makes sense to use the memory of skipper-ingress nodes which are big compared to redis nodes and do not really use much memory. Technically it makes sense to do this in skipper code rather than in redis with lua script or commands. Of course there are a lot of algorithms like expiration already in redis that cry for re-use.
If you would ask me, I would try to be as close to varnish as possible (does not mean redis or not redis), because that's likely the best OSS cache server available and they are around for a while so we can learn from them.
